### PR TITLE
Port all Moreh OPs to compute_output_specs

### PR DESCRIPTION
**Pull request by sminakov-tt**
### Problem description
We're continuing to port all OPs from compute_output_shapes to the new compute_output_specs

### What's changed
Ported all Moreh OPs to compute_output_specs

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/12408552437)
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) tests passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp [hunk 1]
```diff
@@ -62,57 +62,49 @@ void MorehAdamOperation::validate_on_program_cache_hit(
...
+            ret.push_back(out_spec);
+        }
+    }
+    if (tensor_args.output_tensors.at(3).has_value()) {
+        ret.push_back(tensor_args.output_tensors.at(3)->get_tensor_spec());
+    } else if (operation_attributes.amsgrad) {
+        ret.push_back(out_spec);
+    } else {
+        ret.push_back(std::nullopt);
+    }
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp:86]: @sminakov-tt , I am confused, whats going on here?

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp:86]: it was not accurate before?

File: ttnn/cpp/ttnn/operations/moreh/moreh_matmul/device/moreh_matmul_device_operation.cpp [hunk 2]
```diff
@@ -102,78 +102,22 @@ void MorehMatmulOperation::validate_on_program_cache_miss(
     validate_inputs(operation_attributes, tensor_args);
 }
 
-MorehMatmulOperation::shape_return_value_t compute_output_shapes(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_matmul/device/moreh_matmul_device_operation.cpp:105]: interesting :)

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/moreh/moreh_matmul/device/moreh_matmul_device_operation.cpp:105]: need clang tidy for unused functions

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_backward/device/moreh_nll_loss_backward_device_operation.cpp [hunk 3]
```diff
@@ -85,12 +85,15 @@ void MorehNllLossBackwardDeviceOperation::validate_on_program_cache_hit(
...
-    // do not contain the channel_size information.
-    TT_FATAL(false, "moreh_nll_loss_backward not support create output tensors.");
-    return tensor_args.target_tensor.get_shape();
+    if (tensor_args.input_grad_tensor.has_value()) {
+        return {tensor_args.input_grad_tensor->get_tensor_spec()};
+    }
+    return TensorSpec(
+        tensor_args.target_tensor.get_logical_shape(),
+        TensorLayout(
+            tensor_args.target_tensor.get_dtype(), PageConfig(Layout::TILE), operation_attributes.memory_config));
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_backward/device/moreh_nll_loss_backward_device_operation.cpp:96]: To my memory, the `input` and `input_grad` have a shape of `[N, C, d1, d2, ...]`, while `target` has a shape of `[N, d1, d2, ...]`, so returning the shape of the `target` seems incorrect.

If you want to return `input_grad`, you need to take `C` as a separate input.

**Review by hschoi4448 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp (hunk 1 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp:86]: Yes, previously it was inconsistent with  create_output_tensors

File: ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_backward/device/moreh_nll_loss_backward_device_operation.cpp (hunk 3 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_backward/device/moreh_nll_loss_backward_device_operation.cpp]: Thank you for the clarification, I restored the original behavior.

File: ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_unreduced_backward/device/moreh_nll_loss_unreduced_backward_device_operation.cpp [hunk 4]
```diff
@@ -83,13 +83,16 @@ void MorehNllLossUnreducedBackwardDeviceOperation::validate_on_program_cache_hit
...
-    // do not contain the channel_size information.
-    TT_FATAL(false, "moreh_nll_loss_unreduced_backward not support create output tensors.");
-    return tensor_args.target_tensor.get_shape();
+    if (tensor_args.input_grad_tensor.has_value()) {
+        return {tensor_args.input_grad_tensor->get_tensor_spec()};
+    }
+    return TensorSpec(
+        tensor_args.target_tensor.get_logical_shape(),
+        TensorLayout(
+            tensor_args.target_tensor.get_dtype(), PageConfig(Layout::TILE), operation_attributes.memory_config));
```
[the remark this answers is not in the document]
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss_unreduced_backward/device/moreh_nll_loss_unreduced_backward_device_operation.cpp]: Same as above

**Pull request merged**